X509 again

smklein opened the pull request:
Another attempt at https://github.com/oxidecomputer/omicron/pull/1287

In addition to launching an HTTPS server, this also launches an HTTP server so we can smoothly migrate clients (like the CLI).

smklein commented on package-manifest.toml:
```diff
@@ -30,6 +30,14 @@ to = "/var/svc/manifest/site/nexus"
 [[package.omicron-nexus.paths]]
 from = "out/console-assets"
 to = "/var/nexus/static"
+# Note, we could just map the whole "out/certs" directory, but this ensures
+# both files exist.
+[[package.omicron-nexus.paths]]
+from = "out/certs/cert.pem"
+to = "/var/nexus/certs/cert.pem"
+[[package.omicron-nexus.paths]]
+from = "out/certs/key.pem"
+to = "/var/nexus/certs/key.pem"
```
This might be a longer-term issue to resolve, but I'd like for this to not necessarily be bundled with the package.

However, this may require a different mechanism for acquiring / rotation of the certificate.

smklein replied:
I followed up today and met with Rick to discuss this.

My takeaways:
- RSS *must* provide a way for users to input the Nexus IP address + initial certificates
  - Later, an interface through Nexus will support modifying these values
- A key storage mechanism like https://github.com/hashicorp/vault - not necessarily this particular technology, but something highly available - should be used to store keys for both internal and externally facing services.
  - For **internal services**, we could concoct a scheme where:
    - When booted, services generate a public/private key pair
    - The private key stays private to the service while it exists in-memory
    - The public key gets published to "vault" - or whatever key management facility we use - and signed by the internal root CA. This should rely on trusting "who is launching the service" (namely, sled agent) to extend this trust.
  - For **external services**, we rely on user input for these certificates. We could store them in this "vault", but cannot generate them ourselves.

smklein replied:
This is tracked by https://github.com/oxidecomputer/omicron/issues/1528

smklein commented on docs/how-to-run.adoc:
```diff
@@ -53,6 +53,13 @@ This script requires Omicron be uninstalled, e.g., with `pfexec
 
+=== Make me a certificate!
+
+Nexus's external interface will typically be served using public-facing x.509
+certificate. While we are still configuring the mechanism to integrate this real
+certificate into the package system, `./tools/create_self_signed_cert.sh` can be
+used to generate an equivalent self-signed certificate.
```
TODO(me): I wonder if I can fix https://github.com/oxidecomputer/omicron/issues/1398 and make this easier at  the same time?

smklein replied:
See: https://github.com/oxidecomputer/omicron/issues/1528. We're gonna need to grab / insert certs from *somewhere*. 

I've updating the packaging hints to make this more obvious if an error is encountered.

davepacheco commented on common/src/nexus_config.rs:
```diff
@@ -98,16 +98,36 @@ pub enum Database {
     },
 }
 
+/// Describes how ports are selected for dropshot's HTTP servers.
```
I found the naming here a little confusing... what about:

```rust
pub enum ListenPortPolicy {
    HardcodedDefault,
    AnyAvailable
}
```

Even if you don't like those names, does that accurately describe it?  I mean: Nexus isn't _choosing_ based on some algorithm or allocator, is it?  And `Zero` is just a sentinel value used in the sockets API to mean "any available".

davepacheco replied:
Looking at the rest of this, I don't think I follow the motivation for this either.  You could already specify a specific port (including default ports) or "0" to mean "any available", right?

smklein replied:
Are you suggesting I make the configuration file explicitly state the Nexus IP address twice, once for the http interface and once for the https interface?

I was hoping to only have the config file state the necessary IP address once.

I'm happy to change the name of this enum, but it sounds like we should agree whether or not we want it at all first.

davepacheco replied:
> Are you suggesting I make the configuration file explicitly state the Nexus IP address twice, once for the http interface and once for the https interface?

I think that'd be fine and preferable to creating a separate Nexus-level analog for any dropshot config that we want to pass through (now or in the future).  I can see this being a matter of opinion.

smklein replied:
PortPicker removed in the latest iteration; resolving.

davepacheco commented on docs/how-to-run.adoc:
```diff
@@ -53,6 +53,13 @@ This script requires Omicron be uninstalled, e.g., with `pfexec
 that is not the case. The script will then remove the file-based vdevs and the
 VNICs created by `create_virtual_hardware.sh`.
 
+=== Make me a certificate!
```
Is there an analogous change for `how-to-run-simulated`?  In general, the test suite matches what `how-to-run-simulated` does so if you've updated that then things should work.

smklein replied:
There is not, because the implementation within Nexus checks for the existence of the certificate files before deciding whether or not to launch the HTTPS server.

For the simulated server, they won't exist, so only the HTTP server will be launched.

I definitely think this would be required before moving to "HTTPS only", but considered this PR an intermediate step.

davepacheco replied:
Mentioned in the other comment, but that doesn't seem like the right behavior.  It seems like in production, if they don't exist, that should be a fatal error.  And in development, if they do exist, we should start the HTTP server.  If this were part of configuration, then whoever's running Nexus can express their intent.  (I'm not saying we have to do all that in this PR, but if we decide that _is_ what we want, then I think this behavior is not really an intermediate step.)

davepacheco commented on nexus/src/lib.rs:
```diff
@@ -92,26 +96,98 @@ impl Server {
+        // would be the more secure long-term plan, but would make gradual
+        // deployment of this feature more difficult.
+
+        let cert_file = PathBuf::from("/var/nexus/certs/cert.pem");
```
This seems unlikely to work on development machines.

Why not make this part of `dropshot_external`, with `/var/nexus/certs/cert.pem` the value in the SMF config file and something like "out/certs" the value in nexus/examples/config.toml?

smklein replied:
FWIW, this file is only used `if cert_file.exists() && key_file.exists()`.

I was kinda opposed to using the config-file-supplied certificate, because longer-term, it seemed like this value would be:

- Passed from RSS into Nexus: https://github.com/oxidecomputer/omicron/issues/1528 , as part of the rack initialization request 
- Stored in CRDB (and eventually a more secure key store, like Vault): https://github.com/oxidecomputer/omicron/pull/1500/files#r932591199

So although we are currently plumbing the certificates through the package system, I don't expect that to stay the case for very long.

davepacheco replied:
> FWIW, this file is only used if cert_file.exists() && key_file.exists().

What happens if the path is missing on a production system?  It seems like that should produce a fatal error.

> I was kinda opposed to using the config-file-supplied certificate

I can see where you're coming from.  At the same time, it seems like an essential developer or demo tool to be able to start Nexus with a specific certificate?

Is that a good idea for the production config?  I'm not sure.  It might depend on the expected behavior in the longer term.  Is it that Nexus reaches out to CockroachDB/Vault to get the certificate, then loads that into memory, and then starts the HTTPS server with it?  If that's true, then what do we do if Vault is down?  I could see: (1) not coming up at all until we can get the cert, (2) storing the last-known-good certificate to a file and trying to use that until we can get the cert, (3) generating a self-signed certificate and using that until we're able to access the real one.  These all have different tradeoffs.  I've been assuming Nexus could come up (albeit not very usefully) even if everything else in the world was down so that we could at least ask it why it wasn't able to come up.  Anyway, if we go with (2) or (3), then it seems like accepting a file path in the config file would be a reasonable stepping stone.

(Along these lines, I guess hardcoded, deployment-specific absolute paths in source files seem like a code smell to me, even if we gracefully handle them being missing.)

davepacheco reviewed (commented):
Thanks for doing this...it'll be great to have TLS here!

I think all of my comments boil down to the way this gets configured.  I didn't follow why we don't just pass through the corresponding dropshot config.  I may well have missed something here about why that wouldn't work.  But if we could do that, it'd be less code and at the same time preserve more configurability -- some of which seems missing here for the development machine use case.

smklein commented on README.adoc:
```diff
@@ -77,35 +77,15 @@ Supported config properties include:
 |Yes
 |URL identifying the CockroachDB instance(s) to connect to.  CockroachDB is used for all persistent data.
 
-|`dropshot_external`
```
For aforementioned reasons, I'm trying to:

- Take as input an "external IP address"
- Host two dropshot servers, on different ports

Dropshot's configuration expects a SocketAddr, which specifies the IP + port in the configuration file. If I opt to ignore these values (picking 80 and 443 instead, for the port), it felt odd to have a user provide them at all. In lieu of that, I wanted to modify the Nexus configuration to be "precisely the toggles we let a user actually change", no more no less.

Additionally, I found this comment in the Nexus config a little concerning:

```
# Allow larger request bodies (1MiB) to accomodate firewall endpoints (one
# rule is ~500 bytes)
request_body_max_bytes = 1048576
```

I understand why the dropshot config is the way that it is, but if Nexus requires it to be set very particularly to operate correctly, then IMO Nexus should be responsible for setting that config, rather than transparently propagating values which could be inappropriate.

WDYT? This definitely veers into opinion territory.

davepacheco replied:
> For aforementioned reasons, I'm trying to:
> 
>     * Take as input an "external IP address"
> 
>     * Host two dropshot servers, on different ports
> 
> 
> Dropshot's configuration expects a SocketAddr, which specifies the IP + port in the configuration file. If I opt to ignore these values (picking 80 and 443 instead, for the port), it felt odd to have a user provide them at all. In lieu of that, I wanted to modify the Nexus configuration to be "precisely the toggles we let a user actually change", no more no less.

Sorry I didn't see the discussion of why we want to take an external IP address and choose two ports rather than accept two different IP/port pairs.  (Did I miss that?)

I'm gathering there's kind of a philosophical difference here.  My approach with stuff like Nexus is to have configuration be somewhat general-purpose so that it can be customized by developers and test harnesses for various ways of debugging and deploying it.  (I also understand the desire to avoid providing self-destruct buttons, and I think we can still do that.)

> Additionally, I found this comment in the Nexus config a little concerning:
> 
> ```
> # Allow larger request bodies (1MiB) to accomodate firewall endpoints (one
> # rule is ~500 bytes)
> request_body_max_bytes = 1048576
> ```
> 
> I understand why the dropshot config is the way that it is, but if Nexus requires it to be set very particularly to operate correctly, then IMO Nexus should be responsible for setting that config, rather than transparently propagating values which could be inappropriate.

The reason I originally made this configurable was so that we could in principle tune it _up_ in an emergency (as opposed to so that someone trying to tune it down aggressively for some reason).  The challenge is knowing with sufficient confidence that the value selected is big enough.  Now, the config file may not really be the best place for tunables like this in the deployed system, but I think it's the closest thing we've got today to a place where somebody could tune this up in an emergency without rebuilding the software.

smklein replied:
Re-added config - the new implementation allows multiple external configs, so this comment thread can be resolved.

smklein commented:
I've tried to make this more amenable to the idea of "allow the `ConfigDropshot` structure to be plugged directly into Nexus", which attempted to address your comment here: https://github.com/oxidecomputer/omicron/pull/1500#discussion_r938290936

Since Sled Agent is the entity actually responsible for launching Nexus, it is now plumbing this information, by passing a `Vec<ConfigDropshot>` to represent any number of external interfaces. I'm hoping this addresses your comment [here](https://github.com/oxidecomputer/omicron/pull/1500#discussion_r938228360), in that it leaves the input to Nexus "somewhat general-purpose", so developer environments + test harnesses can do whatever they want to do.

I do still think there's a legitimate question of "how do we pass these certificates when they are not part of the Nexus package" - it seems like sled agent will need to load them into the filesystem *somehow* to force them into the Nexus zone, and ultimately the `ConfigDropshot` path, if we are still trying to stick to that configuration.

davepacheco commented on nexus/tests/config.test.toml:
```diff
@@ -41,16 +41,10 @@ max_vpc_ipv4_subnet_prefix = 29
 id = "e6bff1ff-24fb-49dc-a54e-c6a350cd4d6c"
 rack_id = "c19a698f-c6f9-4a17-ae30-20d711b8f7dc"
 
-#
```
Why drop this comment?  It still seems relevant.

smklein replied:
Added it back

davepacheco reviewed (approved):
Awesome!  I like this a lot better.  Thanks.  (I hope you at least don't hate it!)
